model.py

```diff
@@ -462,6 +467,42 @@ def number_of_patrons(self):
         )
         return query.count()
 
+    @classmethod
+    def patron_counts_by_library(self, _db, libraries):
+        """Determine the number of registered Adobe Account IDs
+        (~patrons) for each of the given libraries.
+
+        :param _db: A database connection.
+        :param libraries: A list of Library objects.
+        :return: A dictionary mapping library IDs to patron counts.
+        """
+        # The concept of 'patron count' only makes sense for
+        # production libraries.
+        library_ids = [l.id for l in libraries if l.in_production]
+
+        # Run the SQL query.
+        counts = select(
+            [
+                DelegatedPatronIdentifier.library_id,
+                func.count(DelegatedPatronIdentifier.id)
+            ],
+        ).where(
+            and_(DelegatedPatronIdentifier.type==DelegatedPatronIdentifier.ADOBE_ACCOUNT_ID,
+                 DelegatedPatronIdentifier.library_id.in_(library_ids))
+        ).group_by(
+            DelegatedPatronIdentifier.library_id
+        ).select_from(
+            DelegatedPatronIdentifier
+        )
+        rows = _db.execute(counts)
+
+        # Convert the results to a dictionary.
+        results = dict()
+        for (library_id, count) in rows:
+            results[library_id] = count
+
+        return results
```

Comment:
For a list of 2-tuples turns out you can just call `dict(list_of_2_tuples)`, which I didn't realize till just now. So you could just do `return dict(rows)` here.

There's a small efficiency difference to doing it that way, though it's unlikely to matter at the scale we're working at. This script:

```Python
import random, timeit, uuid

rows = [(str(uuid.uuid4()), random.randint(1000, 100000)) for _ in range(1, 1001)]

def by_for_loop():
  results = {}
  for (library_id, count) in rows:
    results[library_id] = count
  return results

def by_dict_call():
  return dict(rows)

loop_time = timeit.timeit(by_for_loop, number=10000)

dict_time = timeit.timeit(by_dict_call, number=10000)

print("For loop time: ", loop_time)
print("dict() time:   ", dict_time)
```

outputs:

```
For loop time:  0.388839666
dict() time:    0.28115554200000004
```

Reply:
That's good to know.  In this case, though, dict(rows) doesn't work because `rows` is a result set, not an actual list. Converting it to a list requires iterating over it, so I'm not convinced there will be a benefit here. I'll keep this in mind in general though.